Refactor test framework to easily expand on cluster management

smukil opened the pull request:
This patch takes a lot of the code from test/cluster_generator.py
and refactors it into easily manageable classes which allows for
easily creating and destroying a dynomite cluster in a local test
environment. Refer to the last paragraph of this commit message for
the motivation behind this patch.

To create a dynomite cluster, now it's as simple as doing the following:

  cluster = DynoCluster(args.request_file, ips)
  cluster.launch()
  \# Run all the required tests
  cluster.teardown()

This patch does this by forming clear relations between DynoCluster,
DynoSpec, DynoNode and RedisNode.

DynoCluster --manages--> list(DynoSpec), list(DynoNode)
DynoNode    --manages--> RedisNode
DynoNode and DynoSpec have a 1:1 relationship.

In the future, we should decouple RedisNode and DynoNode and abstract
away specific storage node classes to a generic class.
Eg: 'RedisNode' implements 'StorageNode', and 'DynoNode' directly
uses only 'StorageNode'.

This patch also fixes a small bug where the incorrect port is used
for the client listening port.

This patch currently breaks other standalone scripts like func_test.py
and large_value_test.py. This will be cleaned up in a future commit,
and IMO it's not a serious issue as it's difficult to already use
those scripts since they assume that a cluster with a very specific
configuration is already setup (which is not that simple to do
manually today).

The initial motivation behind this patch is to create scripts to
_only start_ or _only stop_ a dynomite cluster for the purpose
of manual testing against a cluster. Before this patch,
cluster_generator.py always cleaned up the resources on exit. Future
commits will introduce new scripts like start_cluster.py and
teardown_cluster.py which will allow developers to just start a cluster
and run tests against it and then tear it down later. Other things
that this patch enables is to do hypothetical experiments like
dual running against 2 dynomite clusters with different
configurations, expanding the framework to run only a subset of tests
based on user flags passed to cluster_generator.py, etc.

Future TODO: Add code comments to classes and methods as necessary.

smukil committed:
Refactor test framework to easily expand on cluster management

This patch takes a lot of the code from test/cluster_generator.py
and refactors it into easily manageable classes which allows for
easily creating and destroying a dynomite cluster in a local test
environment. Refer to the last paragraph of this commit message for
the motivation behind this patch.

To create a dynomite cluster, now it's as simple as doing the following:

  cluster = DynoCluster(args.request_file, ips)
  cluster.launch()
  \# Run all the required tests
  cluster.teardown()

This patch does this by forming clear relations between DynoCluster,
DynoSpec, DynoNode and RedisNode.

DynoCluster --manages--> list(DynoSpec), list(DynoNode)
DynoNode    --manages--> RedisNode
DynoNode and DynoSpec have a 1:1 relationship.

In the future, we should decouple RedisNode and DynoNode and abstract
away specific storage node classes to a generic class.
Eg: 'RedisNode' implements 'StorageNode', and 'DynoNode' directly
uses only 'StorageNode'.

This patch also fixes a small bug where the incorrect port is used
for the client listening port.

This patch currently breaks other standalone scripts like func_test.py
and large_value_test.py. This will be cleaned up in a future commit,
and IMO it's not a serious issue as it's difficult to already use
those scripts since they assume that a cluster with a very specific
configuration is already setup (which is not that simple to do
manually today).

The initial motivation behind this patch is to create scripts to
_only start_ or _only stop_ a dynomite cluster for the purpose
of manual testing against a cluster. Before this patch,
cluster_generator.py always cleaned up the resources on exit. Future
commits will introduce new scripts like start_cluster.py and
teardown_cluster.py which will allow developers to just start a cluster
and run tests against it and then tear it down later. Other things
that this patch enables is to do hypothetical experiments like
dual running against 2 dynomite clusters with different
configurations, expanding the framework to run only a subset of tests
based on user flags passed to cluster_generator.py, etc.

Future TODO: Add code comments to classes and methods as necessary.

hashbrowncipher commented on test/utils.py:
```diff
@@ -1,9 +1,69 @@
+    return sum(count for rack, count in dc)
+
+def generate_ips():
+    state = BASE_IPADDRESS
```
Hmm this could be itertools.count.

(Reviewing my own code, I think)

smukil replied:
Done.

hashbrowncipher commented on test/utils.py:
```diff
@@ -1,9 +1,69 @@
+    """Converts the request into an easy to consume dict format.
+
+    We don't ingest the request in this format originally because dict
+    ordering is nondeterministic, and one of our goals is to deterministically
```
I think this is no longer true in Python 3.6. I wonder if we should adapt the code to newer versions of Python.

smukil replied:
I think I'll hold off on this for now since I don't want to introduce a minimum Python version requirement at this point.

hashbrowncipher commented on test/dyno_node.py:
```diff
@@ -2,16 +2,42 @@
-    def __init__(self, ip, port=8102,
-                 dnode_port=8101, data_store_port=1212):
-        super(DynoNode, self).__init__(ip, port)
+    '''TODO: Decouple from Redis and make it more generic'''
```
This kind of comment is typically prefaced with `#`, not `"""`. `''''` is valid, but rare in Python syntax.

If you want, you could download code formatter like [black](https://pypi.org/project/black/). It will do all this for you.

smukil replied:
Done.

hashbrowncipher commented on test/cluster_generator.py:
```diff
@@ -202,26 +37,40 @@ def main():
+    with local.cwd(temp):
+        try:
+            # Launch the standalone Redis node and the dynomite cluster.
+            standalone_redis.launch()
```
Unless there's a reason not to, I think it's worth preserving this as a context manager API. Let's discuss it in person.

smukil replied:
Thanks for this pointer. Makes the code much cleaner.

smukil committed:
Add scripts to start and stop a test cluster

This path adds start_cluster.py and kill_cluster.py to start and
stop a cluster respectively. Once a cluster is started, a description
of that cluster is written to a YAML file in 'running_cluster.yaml'.

When kill_cluster.py is invoked, the script reads through the YAML
file if it exists, and kills all the PIDs listed.

On every new cluster startup, there is a pre-sanity check to determine
whether a cluster is already running, and if it is, it is torn down
first before starting a new one.

Future TODOs: Add generic utils for writing YAML files instead of
the ad-hoc ways it's done in multiple places now.

smukil committed:
Allow request YAML files for cluster generation to have dynamic configs

This patch allows the test cluster to be started with configuration
values such as 'read_consistency', 'write_consistency', etc.

Currently, the code only supports single value configuration keys.
Multi-value support will be added as necessary.

smukil committed:
Fix incorrect topology bug in test cluster creation

This patch does 3 things:
1) It fixes a longstanding bug in the DynoSpec class that mixed up
   a rack with a DC, causing a cluster of a different topology to
   be created than the one requested.

2) It fixes a small bug introduced in this series of test framework
   commits (in the last 3 commits), which mixed up a client port with
   a internode port.

3) It adds a clear distinction between 'client_port' and 'dnode_port'
   in the DynoSpec class, rather than just having a 'port' field
   which was a source of confusion and led to the above bug (Bug #2).

smukil commented:
I'm merging this to unblock myself. If anyone has more review comments for this in the future, I'll address them in a separate patch.
